examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -170,6 +139,40 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
         latch.countDown();
     }
 
+    private int processMessages(int messageProcessed, ConsumerRecords<Integer, String> records)
+        throws ProducerFencedException, FencedInstanceIdException {
+        try {
+            // Begin a new transaction session.
+            producer.beginTransaction();
+            for (ConsumerRecord<Integer, String> record : records) {
+                // Process the record and send to downstream.
+                ProducerRecord<Integer, String> customizedRecord = transform(record);
+                producer.send(customizedRecord);
+            }
+            Map<TopicPartition, OffsetAndMetadata> positions = new HashMap<>();
+            for (TopicPartition topicPartition : consumer.assignment()) {
+                positions.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
+            }
+            // Checkpoint the progress by sending offsets to group coordinator broker.
+            // Under group mode, we must apply consumer group metadata for proper fencing.
+            if (this.mode.equals("groupMode")) {
+                producer.sendOffsetsToTransaction(positions, consumer.groupMetadata());
+            } else {
+                producer.sendOffsetsToTransaction(positions, consumerGroupId);
+            }
+
+            // Finish the transaction. All sent records should be visible for consumption now.
+            producer.commitTransaction();
+            messageProcessed += records.count();
+        } catch (CommitFailedException e) {
```

Comment:
I don't think this is the only case that we can abort a transaction. This only handles offset commit failures, but what about send failures?